lib/syntax.ml

```diff
@@ -177,6 +187,32 @@ and assign_types expr =
             convert ret, List.map convert args
           else
             fail "types do not match : %s" (show ())
+        | InF (lhs, rhs), _ ->
```

Comment:
this whole branch seems equivalent to above branch applied to `F (Bool, [lhs; Tuple rhs])`
ie either it can be factored better or `InF` can be completely replaced with `F`

Reply:
my first attempt used `F (Bool, _)` but I couldn't get type unification to work, will try again